consensus/src/runway/mod.rs

```diff
@@ -264,40 +321,6 @@ where
         }
     }
 
-    // TODO: we should return an error and handle it outside
-    fn validate_unit(
-        &self,
-        uu: UncheckedSignedUnit<H, D, MK::Signature>,
-    ) -> Option<SignedUnit<'a, H, D, MK>> {
-        let su = match uu.check(self.keybox) {
-            Ok(su) => su,
-            Err(uu) => {
-                warn!(target: "AlephBFT-runway", "{:?} Wrong signature received {:?}.", self.index(), &uu);
-                return None;
-            }
-        };
-        let full_unit = su.as_signable();
-        if full_unit.session_id() != self.session_id {
-            // NOTE: this implies malicious behavior as the unit's session_id
-            // is incompatible with session_id of the message it arrived in.
-            warn!(target: "AlephBFT-runway", "{:?} A unit with incorrect session_id! {:?}", self.index(), full_unit);
-            return None;
-        }
-        if full_unit.round() > self.store.limit_per_node() {
-            warn!(target: "AlephBFT-runway", "{:?} A unit with too high round {}! {:?}", self.index(), full_unit.round(), full_unit);
-            return None;
-        }
-        if full_unit.creator().0 >= self.n_members.0 {
```

Comment:
Are we relying here on the fact that the underlying implementation of `KeyBox` in `uu.check(...)` is correctly dealing with incorrect node indices? 

No matter whether this is the case I think it's way better to have an explicit error variant `IncorrectNodeIndex` instead of relying on the signature check.

Reply:
Yes, that's what I am relying on. I removed it when I noticed there is no way of testing this, since it cannot happen by design. Because of that I don't think we should put it back.

Reply:
I don't get it -- what "cannot happen by design"? 

I think that the `IncorrectNodeIndex` is semantically very different to `WrongSignature` and it is not quite intuitive for the latter to capture the former.

Reply:
And what I would suggest is to move the check for correctness of the `NodeIndex` before the signature check, because without this there is indeed no point.

Reply:
But it always will, so this part of the code will never be reached. Dead code ain't good.

Reply:
Ok, I see, the problem is that the whole `Signed` design is not allowing us to reach the NodeIndex before validating the signature?

Reply:
Ah, async communication, I haven't seen your other comment before writing my own so my response was a non sequitur, classic.

We could actually check it in this order, `UncheckedSigned` gets you access to the signed object. It is only a) doubled with the signature check and b) out of order with all the other checks. Currently, we first check the signature and only afterwards the shape of the unit, going from outside in. Doing this one check out of order to check something that will be re-checked in a moment anyway seems wrong.

Perhaps it would be nice to make this a more explicit part of the `KeyBox` interface, maybe even with different errors returned from `check`ing. `KeyBox` could also keep and make accessible the information about the node count, which would simplify some other parts of the code. It already has to have that information implicitly, so explicitly requiring it would solve this problem. What do you think about this idea? (Not as part of this PR, but as a task for the future.)

Reply:
Yeah, the proposal sounds good. But one thing that I think begs to be done is to at least write in the doc comment for the `KeyBox` trait that the implementation is expected to return `false` when the index is incorrect.